### ISSUE TITLE
Application is locked - reauthentication required 

### ISSUE THREAD

**Issue by porta-logica** [IbcAlpha/IBC#277]
I run IBC in a docker container. Occasionally, due to reasons unknown to me, the TWS application (live) becomes locked. In this case a pop-up window appears on the screen with the caption: "Enter trader workstation password to unlock the application". Can IBC capture this event and put the password in the field?

**Comment by rlktradewright** [comment]
Have you set 'Never lock Trader workstation' in the Lock and Exit section of the TWS/Gateway configuration page?

**Comment by porta-logica** [comment]
IBKR has removed that option in the TWS version 10.16.1.i (Dec 2023). Downgrade seems a poor workaround to me.

**Comment by rlktradewright** [comment]
There have been varioous reports about this - for example see #257, but it seems to come and go. There have also been suggestions that this is particularly the case for Hong Kong users, and that the HK authorities are demanding this auto-lock after a period of time.

Are you an HK user? 

I notice that you're on a rather old version of TWS. Be aware that the only versions that have bug fixes applied are the 'latest' version. and the current 'stable' version. So if for example IBKR put this right in TWS 10.18, it will never be fixed in 10.16.

Therefore I advise you to upgrade your TWS to the current stable version, and see if the problem still exists. If it does, the next thing you should is submit a ticket to IBKR asking what the reason for this limitation is.

If they say that this is a requirement of the Financial Trading authorities, then i'd have to think very carefully about whether I should amend IBC to automatically re-authenticate: I don't want to be implicated in law-breaking!

It should be perfectly possible for IBC to do this, but I'm reluctant to do it until we know what the official story is.

**Comment by porta-logica** [comment]
Thanks for your comment. Please review my findings which may answer some of your questions.

1. I use the latest build:

Trader Workstation Copyright © 2000-2024 LICENSOR to Interactive Brokers ALL RIGHTS RESERVED
Build 10.32.1n, Dec 3, 2024 3:53:47 PM Jolt
Build 1.18.12, Sep 5, 2024 06:57:46 PM Nia
Build 2.25.7, Sep 24, 2024 06:08:56 PM ModelNav
Build 1.13.2, Jan 24, 2022 04:59:21 PM Riskfeed
Build 2.45.14, Oct 22, 2024 04:35:57 PM
Java Version: 1.8.0_202, OS: Linux (amd64, 6.8.0-49-generic),
Memory: heap max 2048Mb, current 199Mb
This software is covered by U.S. Patent No. 7.827.092.

2. I am a not a HK user. I have registered in Switzerland.

3. The issue is independent of the client geographic location. I have tested on clients in the EU and USA as well.

4. The issue occurs only in applications authenticated with *live* account credentials. Paper account connections are *not* interrupted, even in a concurrent connection.

5. https://github.com/wibuf has already reported the issue in a comment https://github.com/IbcAlpha/IBC/issues/257#issuecomment-2214763446 to https://github.com/IbcAlpha/IBC/issues/257. That is also my experience. I would rephrase his concern: the unhandled lock condition renders IBC unusable for live trading, degrading it to a paper trading environment.

6. In contrast to IBC, the official TWS application, JVM on Linux, has not been interrupted for several days using a live or paper account.

7. For this reason, a ticket to the IBKR support could only include questions about IBC which would be awkward. I have no idea, how their code can tell the difference between IBC and TWS pure; I have guesses, though. I also cannot exclude the possibility that the lock behavior was implemented on response to a regulatory requirement.

**Comment by rlktradewright** [comment]
I get the impression from your points 6 and 7 that you harbour a suspicion that this problem may actually be somhow caused by IBC.

I can categorically state that this is not the case. The problem lies with TWS, and it's an issue for IBKR. The only reason I'm engaging in this thread at all (which has consumed quite a lot of my valuable time) is that I do actually care about people getting the best out of the combination of TWS and IBC, as I do my self, and presumably the vast majority of the IBC users do, since only a handful of users seem to have been affected by this.

Let me explain how I can robustly make this assertion. What IBC does, running in the same process as TWS, is set up a listener using standard Java mechanisms which is notified of events occuring in the user interface, in particular events to do with JFrames and JDialogs (ie forms and dialog boxes) and button clicks: for example when a form is loaded or unloaded or gets the focus. 

Most such events are ignored, because they are of no interest to IBC. But if, for example, the TWS login dialog is displayed, IBC inserts the username and password and clicks the login button, behaving exactly like the user to initiate the authentication process. 

There is nothing that IBC does that cannot be done by the user, and IBC never changes the content or structure of any UI element, except for making valid user input into fields, checkboxes, combo boxes etc. So if your configuration dialog does not contain the 'Never lock Trader Workstation' checkbox then it can only be because TWS displayed it like that: there is no way for IBC to remove that element (theoretically it could, but as I said, IBC doesn't modify the user interface structure, ever).

My problem with all this is that I have zero eveidence to go on, apart from suggestions that this started as a response to a change in HK regulations, and the various screenshots and descriptions that a small handful of users have submitted. Clearly there _is_ a problem somewhere, but I'm as completely in the dark about this as anyone else.

My suspicion is that IBKR are trying to find a way to comply with this HK ruling, without impacting users who are not subject to it, but they haven't got it poperly working yet, and occasaionally it sort of leaks out and affects some others. I have certainly never experienced anything like this, and I run TWS with IBC all day every day of the week (with the necessary daioly and weekly restarts).

What might be helpful would be if people like yourself who do experience the problem could do some systematic investigation. For example:

* Does this happen every day, multiple times a day, once a week, random intervals or what?
* Is it always the same inactivity timeout?
* What things count as 'activity'?

As far as using IBC to mitigate the problem is concerned, the most effective thing might be for IBC to do something periodically that counts as activity, so that the inactivity timeout is never activated. This would be better than automatically re-authenticating when TWS is locked, because having it locked at all might play havoc with an auto-trading system. But as I've said before, I don't want to get into doing this without IBKR being made aware of the issue so that they can actually fix it. 

So someone needs to submit a fault report including the TWS logfile, asking why the 'Never lock Trader Workstation' checkbox is missing.

**Comment by porta-logica** [comment]
Sorry for having made you feel that I somehow blame you or IBC. My bad. It is definitively not my intention.

However your impression does not follow from my findings. You eloquently disprove something I did not claim.

I claim that the issue is caused by an **intentional IBKR feature** which is manifested in an IBC environment. It is not *caused* by IBC. However it is not a bug in IBKR, either.

Unfortunately this very IBKR feature goes against the declared aim of IBC and breaks its usability in live environments.

I also doubt that the issue has something to do with "inactivity" which is a concept to be clarified. That is stated above in p.6. I also doubt that the issue is relevant only for HK users, see p.2.

Please find attached a screenshot of the pop-up. It appears in about 30 minutes after (re-)authentication in a predictable pattern.

![Unlock Application Dialog](https://github.com/user-attachments/assets/ce70a926-5e80-4873-9963-10c808239063)

It may be true that only a handful of users are affected. I just wanted you to increase that number by one.

**Comment by rlktradewright** [comment]
_[This is quite a detailed post, so if you don't have time to read it, you can just skip to the section entitled HOW TO FIX THIS at the end, and take it from there.]_

It's not a question of 'blame'. and there's nothing you need to apologise about. 

Of course it's an intentional IBKR feature (that's what I said: IBC can't cause this, ergo it must be IBKR, and it presum,ably happens because of something that TWS responds to).

But you can't conclude from the indisputable fact that it manifests in an IBC environment for you, that therefore it _only_ manifests in an IBC environment for everyone. This shows that you are not familiar with the full history of this phenomenon.

So take a look at this thread on the TWS API User Group: https://groups.io/g/twsapi/topic/tws_option_never_lock_trader/103503025

and this one on the IBC User Group::
https://groups.io/g/ibcalpha/topic/106239286#msg2242

You should now understand where the HK angle comes from.

I don't see why you think there's something unclear about 'inactivity': it presumably means that nothing has interacted with TWS during the specified inactivity timeout, and therefore TWS will be locked, on the assumption that the user is not currently actively using it. It's true that what actually constitues 'activity' is not specifically defined, so it's not obvious whether moving the mouse pointer across the TWS window, without clicking anything, counts as 'activity' (I would expect that it isn't).

Ok, let's take stock of the evidence that we have so far:

* IBKR produced a version of TWS that did not show the 'Never lock Trader Workstation' checkbox
* It appears that this version was produced in response to a regulatory change, as stated in the relevant TWS Release Notes:
```
IB-HK Account Users Only
To meet Compliance requirements, users with an IB-HK account will not be able to disable
the TWS 30 minute session timeout. For other entities, this feature can be managed via Global
Configuration from the Lock & Exit page.
```
* This applied regardless of whether IBC was in use.
* This behaviour appears to have 'leaked' to non-HK users.
* In some cases upgrading to a later TWS version seems to have corrected the fault for non-HK users
* in other cases, it seems to have persisted through multiple TWS versions
* It's not clear whether this still happens for HK users: no-one is saying much, so I suspect that the behaviour has been quietly shelved.  Either that, or the HK users have learned to live with it somehow, perhaps by setting the inactivity timeout to be long enough to get them to the next auto-restart and therefore never getting locked

I think this is enough to piece together a reasonable conjecture, and to propose a possible solution: I'll explain.

1. IBKR produced an initial version, in a 'latest' TWS.
2. A small proportion of TWS users installed that version.
3. There was a bug which resulted in (some?) non-HK users also experiencing loss of the 'Never lock Trader Workstation' checkbox.
4. TWS wrote something (let's call it 'X') to its settings file which inicated that the 'Never lock Trader Workstation' checkbox should not be displayed.
5. If a user explicitly uninstalled this errorneous version, X was removed from the settings file. So installing a new version that didn't contain the flaw would make everything fine again.
6. If a user just installed a new version without explcitly removing the faulty version, X remained in the settings file, so installing a new version would not help.

I think this explains all the observations. In your case, you don't experience the missing checkbox in your paper-trading configuration because either you never installed the faulty version, or if you did you explicitly uninsstalled it.

For your live configuration, you did install it, but you didn't explicitly uninstall it, and so X still remains in your TWS settings, so you don't see the checkbox.

Regarding your point 6 in your earlier post: running TWS without using IBC probably stores its settings in a different TWS location - this is definitely the case unless you have explicitly configured IBC and TWS to use the same location. So if you never ran TWS in this way while you had the faulty version installed, you will never have had X in that settings file, so it will all work correctly.

### HOW TO FIX THIS

If the description above is correct, then to fix it we need to remove X from the relevant TWS settings file. Since we don't know what or where X actually is in the settings file, and anyway the settings file is encrypted so it can't viably be changed, the only thing we can do is to delete the settings file, so that TWS will re-create it when it next startes. Assuning your settings are also stored on the IBKR servers then it should automatically pick up all your tickers, watchlists etc. With any luck, X is not also stored on the server, and you will end up with a 'clean' settings file and all will be well.

To determine where your settings file is, look at any IBC logfile for your live IBC/TWS setup, and find the entry that contains this:

`IBC: TWS Settings directory is: `

That directory will contain a subdirectory with a long name similar to `cefiodafddealakeckhkohbbaenmdfncgjgmpgpb`, and you can just delete this directory in its entirety (or rename it so you can reinstate it if you need to for some reason).

Then restart IBC, and with any luck the problem will be solved.

**Comment by mlvn23** [comment]
It looks like all my IB instances are now like this. I didn't change anything for the past 6 months or so. What's more, my app cannot connect to TWS if the application is locked. So it's going to be an issue if the connection is lost.

**Comment by rlktradewright** [comment]
I think IBKR did something silly that affected all TWS instances. But if you restart them, they seem to be ok now, so they must have realised their mistake and fixed it.

**Comment by fuzihaofzh** [comment]
I am not registered in HK but also see this problem. It still exists until now.

**Comment by rlktradewright** [comment]
@fuzihaofzh 
Well, report it to IBKR - perhaps they can tell you what the cause is. There is nothing IBC can do about it. See my post above on Dec 10 2024.
